File: app/src/main/java/org/pw/masterthesis/peaksrecognition/ScreenManager.java
package org.pw.masterthesis.peaksrecognition;

import android.opengl.GLU;

public class ScreenManager {
    private final int[] viewport;
    private float[] viewMatrix;
    private float[] projectionMatrix;

    public ScreenManager(int screenWidth, int screenHeight) {
        viewport = new int[]{0, 0, screenWidth, screenHeight};
    }

    public void setMVPMatrices(float[] viewMatrix, float[] projectionMatrix) {
        this.viewMatrix = viewMatrix;
        this.projectionMatrix = projectionMatrix;
    }

    public float[] getScreenPosition(float xVertex, float yVertex, float zVertex) {
        float[] screenPosition = new float[3];
        GLU.gluProject(xVertex, yVertex, zVertex, viewMatrix, 0, projectionMatrix, 0, viewport, 0, screenPosition, 0);

        return screenPosition;
    }

    public boolean checkIfPointOnScreen(float[] screenPoint) {
        return screenPoint[0] >= 0 && screenPoint[0] <= viewport[2]
                && screenPoint[1] >= 0 && screenPoint[1] <= viewport[3]
                && screenPoint[2] >= 0 && screenPoint[2] <= 1;
    }
}

File: project/python_prototype/ScreenManager.py
import numpy as np
from OpenGL.GLU import *


class ScreenManager:
    def __init__(self, width, height):
        self.viewport = np.array([0, 0, width, height])
        self.view_matrix = None
        self.projection_matrix = None

    def set_MVP_matrices(self, view_matrix, projection_matrix):
        self.view_matrix = view_matrix
        self.projection_matrix = projection_matrix

    def get_screen_point(self, vertex_x, vertex_y, vertex_z):
        screen_position = gluProject(vertex_x, vertex_y, vertex_z, self.view_matrix, self.projection_matrix,
                                     self.viewport)
        return screen_position

    def check_if_point_on_screen(self, screen_x, screen_y, screen_z):
        return ((0 <= screen_x <= self.viewport[2])
                and (0 <= screen_y <= self.viewport[3])
                and (0 <= screen_z <= 1))
